DylibBinder/DBGenericParameter.cs

```diff
@@ -0,0 +1,155 @@
+		}
+
+		public int Depth { get; }
+		public int Index { get; }
+		public string Name { get; }
+
+		public bool Equals (DBGenericParameter other)
+		{
+			if (this.Depth == other.Depth && this.Index == other.Index)
+				return true;
+			return false;
+		}
+	}
+
+	internal class DBGenericParameters {
+
+		public DBGenericParameters (SwiftBaseFunctionType signature) {
+			AssignGenericParameters (signature);
+		}
+
+		public DBGenericParameters (SwiftType swiftType)
+		{
+			FilterGenericParameterSource (swiftType);
+		}
+
+		public DBGenericParameters (DBTypeDeclaration typeDeclaration)
+		{
+			ParseTopLevelGenerics (typeDeclaration.Funcs, typeDeclaration.Properties);
+		}
+
+		public List<DBGenericParameter> GenericParameters { get; } = new List<DBGenericParameter> ();
+
+
+		void AssignGenericParameters (SwiftBaseFunctionType signature)
+		{
+			FilterGenericParameterSource (signature.Parameters, signature.ReturnType);
+		}
+
+		void FilterGenericParameterSource (params SwiftType [] types)
+		{
+			foreach (var type in types) {
+				switch (type) {
+				case SwiftFunctionType funcType:
+					HandleFunctionType (funcType);
+					return;
+				case SwiftBoundGenericType boundType:
+					HandleBoundGenericType (boundType);
+					return;
+				case SwiftTupleType tupleType:
+					HandleTupleType (tupleType);
+					return;
+				case SwiftGenericArgReferenceType refType:
+					HandleGenericArgReferenceType (refType);
+					return;
+				default:
+					return;
+				}
+			}
+		}
+
+		void HandleFunctionType (SwiftFunctionType funcType)
+		{
+			switch (funcType.Parameters) {
+			case SwiftBoundGenericType boundType:
+				HandleBoundGenericType (boundType);
+				return;
+			case SwiftTupleType tupleType:
+				HandleTupleType (tupleType);
+				return;
+			case SwiftGenericArgReferenceType refType:
+				HandleGenericArgReferenceType (refType);
+				return;
+			default:
+				return;
+			}
+		}
+
+		void HandleBoundGenericType (SwiftBoundGenericType boundType)
+		{
+			foreach (var bound in boundType.BoundTypes) {
+				if (bound is SwiftGenericArgReferenceType refType)
+					HandleGenericArgReferenceType (refType);
+			}
+		}
+
+		void HandleTupleType (SwiftTupleType tupleType)
+		{
+			foreach (var content in tupleType.Contents) {
+				switch (content) {
+				case SwiftBoundGenericType boundGType:
+					HandleBoundGenericType (boundGType);
+					return;
+				case SwiftFunctionType cFuncType:
+					HandleFunctionType (cFuncType);
+					return;
+				case SwiftGenericArgReferenceType refType:
+					HandleGenericArgReferenceType (refType);
+					return;
+				case SwiftTupleType innerTupleType:
+					HandleTupleType (innerTupleType);
+					return;
+				default:
+					return;
+				}
+			}
+		}
+
+		void HandleGenericArgReferenceType (SwiftGenericArgReferenceType refType)
+		{
+			AddIfNotPresent (new DBGenericParameter (refType.Depth, refType.Index));
+		}
+
+		void AddIfNotPresent (DBGenericParameter newGP)
```

Comment:
The BCL has collections that are implemented to allow the addition of objects it not already present.